store/store.go

```diff
@@ -183,15 +223,20 @@ func NewQueuedRunChannel() RunChannel {
 }
 
 // Send adds another entry to the queue of runs.
-func (rq *QueuedRunChannel) Send(rr models.RunResult, ibn *models.IndexableBlockNumber) error {
+func (rq *QueuedRunChannel) Send(jobRunID string, rr models.RunResult, ibn *models.IndexableBlockNumber) error {
 	rq.mutex.Lock()
 	defer rq.mutex.Unlock()
 
 	if rq.closed {
 		return errors.New("QueuedRunChannel.Add: cannot add to a closed QueuedRunChannel")
 	}
 
+	if jobRunID == "" {
+		return errors.New("QueuedRunChannel.Add: cannot add an empty jobRunID")
```

Comment:
Thought about this overnight and came to the following, somewhat longwinded conclusion. Stay with me 😅please.

*I \*think\* this is technically a precondition rather than a program invariant due to its scope being the method and not the lifetime of an instance, class, or program. Let me know if I'm wrong, but I'll put semantics aside and continue with the core question being, should we `panic` due to something we expect to always be true?*

-----

While it may be an invariant, I don't think that naturally assumes we should `panic` in its event on production. While we hope it never happens, it would be a shame to take down a node and all other runs due to one isolated issue. Had this been the case earlier, our users would have been blocked and had an unusable binary. Perhaps we could `recover` from said `panic`, but at that point I would rather return the error as we are doing now.

There are many discussions around how to handle preconditions and program invariants. I lean towards the model that `asserts` in development/debug, but logs/noops in production. Especially in our scenario, since it's isolated to a specific job run.

I didn't find as much material as I thought I would on this. Maybe the practice of forcing preconditions or invariants is no longer in vogue. I didn't want to bring in resources regarding design by contracts, as that would take me even farther down the rabbit hole. Swift has something simple here:

| function     | debug | release |
|--------------|-------|---------|
| assert       | YES   | NO      |
| precondition | YES   | YES     |

ref: https://stackoverflow.com/a/29673504/639773

Java has the `assert` keyword that usually becomes a noop in production:
https://en.wikibooks.org/wiki/Java_Programming/Keywords/assert

The assert model above (noop in prod, not to be confused with go's test assert) is what I would conceptually employ here. But I prefer our current practice of following go's idioms by returning an error, regardless of the environment, and handling it as we see fit. In this case logging an error, and keeping the node up.

If we do go the `panic` route, would we want to set a precedent that all preconditions `panic` on failure? I think no, because that would lead to catastrophic failures on preconditions we suspect to be true but can always discover otherwise when it's too late and we're already running.

Reply:
Ultimately the line of thinking that leads to using a panic, in production code is thus: have you reasoned about what the program is capable of in the event that the invariant is violated? Because if you haven't, your software effectively has a license to do whatever it wants. It is therefore better to fail fast and quickly and be made acutely aware of the bug in your code and have a chance to fix it, then let your code silently continue and secretly launch the entire nuclear arsenal of the United States ;)

Anyway, how about your suggested location below?

Reply:
Regarding the location below, I'll go ahead and change that to a `panic`.